dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,246 @@
+ */
+template <typename Backend>
+kernels::TensorListShape<> PropagateShapes(ExpressionBase &expr,
+                                           const workspace_t<Backend> &ws) {
+  // TODO(klecki): assume that everything is binary for now
+  if (expr.IsLeaf()) {
+    if (!expr.IsTensor()) {
+      expr.shape = kernels::TensorListShape<>{{1}};
+    }
+    expr.shape = ws.template InputRef<Backend>(expr.GetMappedInput()).shape();
+    return expr.shape;
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_shape = PropagateShapes<Backend>(expr[0], ws);
+    auto right_shape = PropagateShapes<Backend>(expr[1], ws);
+    expr.shape = ShapePromotion(expr.GetOp(), left_shape, right_shape);
+    return expr.shape;
+  }
+  DALI_FAIL("Only binary expressions are supported");
+}
+
+template <typename Backend>
+void SelectExpressionImpl(ExpressionBase &expr) {
+  if (expr.IsLeaf())
+    return;
+  expr.impl = ExprImplFactory(workspace_t<Backend>{}, expr);
+  for (int i = 0; i < expr.GetSubexpressionCount(); i++) {
+    SelectExpressionImpl<Backend>(expr[i]);
+  }
+}
+
+template <typename Backend>
+class ArithmeticGenericOp : public Operator<Backend> {
+ public:
+  inline explicit ArithmeticGenericOp(const OpSpec &spec) : Operator<Backend>(spec) {
+    // partial specification, needs to be filled in
+    // expr_ = Expression(spec.GetArgument<std::string>("expression_desc"));
+    expr_ = ParseExpressionString(spec.GetArgument<std::string>("expression_desc"));
+    num_tasks_ = 2 * num_threads_;
+  }
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+
+    result_type_id_ = PropagateTypes<Backend>(*expr_, ws); // can be done once
+    exec_order_ = CreateExecutionOrder(*expr_); // can be done once
+    SelectExpressionImpl<Backend>(*expr_); // can be done once ?? or after the shape part<<>>>>
+
+    // TODO - we may need to set the shapes of intermediate results from bottom-up
+    result_shape_ = PropagateShapes<Backend>(*expr_, ws); // need to be done every iteration
+    AllocateIntermediateNodes(); // do it after the shapes are propagated
+
+    output_desc[0] = {result_shape_, TypeTable::GetTypeInfo(result_type_id_)};
+    sample_cover_ = GetCover(result_shape_, num_tasks_);
+    task_cover_start_ = GetCoverStarts(sample_cover_, num_tasks_);
+
+    return true;
+  }
+
+  using Operator<Backend>::RunImpl;
+  void RunImpl(workspace_t<Backend> &ws) override;
+
+ private:
+  void AllocateIntermediateNodes() {
+    auto &expr = *expr_;
+    bool is_simple_expression = expr.GetSubexpressionCount() == 2 && expr[0].IsLeaf() && expr[1].IsLeaf();
+    DALI_ENFORCE(is_simple_expression, "Complex expression trees are not supported");
+    //TODO: allocate memory for intermediate results and point the threads to them
+    //it should be possible to be done once, as types and shape of the expression
```

Comment:
> and shape
That's simply not true...

Reply:
Yes, I didn't take the support/scalar case into account. 
If we were to distinguish only between scalar vs tensor it would hold (not exactly a shape, but that's what I had in mind).

Reply:
> not exactly a shape

^^ that